### ISSUE TITLE
Any existing examples of generator to use with -lazy ?

### ISSUE THREAD

**Issue by Kimserey** [tsenart/vegeta#353]
#### Question

I would like to generate random body for each POST requests executed. I read the [post from thisdata.com](https://thisdata.com/blog/load-testing-api-interfaces-with-go-and-vegeta/) and the issues linked to that _random body_ scenario and saw that it all pointed to having a *generator providing a stream of input* to vegeta using `-lazy` option.

I understand that it can be written in any language/script but I am having a hard time to understand how to write one and even find an example of a _generator_ that could be piped to vegeta. 

Would you have an existing example?

**Comment by tsenart** [comment]
Perhaps you’d find the new JSON target format easier to produce?

**Comment by Kimserey** [comment]
Thanks for the reply @tsenart , yes JSON target format will definitely make it easier.

But I am still trying to understand how does `-lazy` works. For example if I make a script that prints every 10 seconds a JSON formatted target for a period of 1 minute, does that mean that `vegeta` will attack every 10 seconds for a duration of 1 minute?

Would the following be the right command?

```
> my_script_generator.sh | vegeta -lazy
```

**Comment by tsenart** [comment]
Your script should produce targets as fast as it can with no sleeping in between targets.

**Comment by Kimserey** [comment]
> Your script should produce targets as fast as it can with no sleeping in between targets.
Got it, and `vegeta` would select the latest target generated?

**Comment by tsenart** [comment]
No. Your generator blocks if vegeta attack is too slow to consume targets. On the other hand, if vegeta attack is too fast, it won’t be able to sustain your desired request rate. Generated targets are not dropped. By the way, if you use CTRL-C to stop the attack, it’s normal to see a broken pipe error; basically it’s your generator that is trying to write to a closed pipe, because vegeta already quit.

**Comment by Kimserey** [comment]
> Your generator blocks if vegeta attack is too slow to consume targets. On the other hand, if vegeta attack is too fast, it won’t be able to sustain your desired request rate.
Oh okay, my bad this was due to my misunderstanding of how pipe worked.

So if I were to create a test for a duration of 10 minutes, the requirements for the generator would be:

1. generate value faster than the rate
2. generate for a duration of 10 minutes

Otherwise the command `my_script_generator.sh | vegeta -lazy -duration=10m` will never stop right?

**Comment by tsenart** [comment]
Yes to point number 1. The second isn’t needed. vegeta attack will stop after 10 minutes and your script will fail to write to stdout with the broken pipe error.

**Comment by Kimserey** [comment]
It doesn't seem to stop I am on Windows running `vegeta` v12.1.0. It will wait until my generator completes. If the generator never completes, I have to CTRL+C.

But this question can be closed, as I understand now how to get `-lazy` to work.

**Comment by Kimserey** [comment]
Just to clarify a bit further, the generator wouldn't need run for a special duration - it would only need to ensure the rate and the amount of targets to be generated.

For example if I were to run a test for `10 rps` for a duration of `1 min`, the requirements for the generator would be that it needs to generate `600 targets (10 rps * 60 sec)`.

**Comment by tsenart** [comment]
Vegeta reads targets on demand when -lazy=true. You can use buffered IO to stdout or not. If you don’t, then your script will block waiting between each target read.
